### DISCUSSION TITLE
isowords app structure

### DISCUSSION THREAD

**Discussion by gerfarfel** [pointfreeco/episode-code-samples#75] [Q&A] [answered]
I have a question relating to episode 134, Concise Forms: The Point. I'm interested in understanding how the isowords app is structured. It appears that there's the isowords app itself, and also a separate swift package, also named isowords, containing a bunch of modules. Specifically, I'm interested in:

• What does the directory structure of the project look like? I.e., what's the relationship between the app and the Swift package?

• What's the procedure for adding a new module to the isowords package? I.e., what are the exact steps when you want to add a module?

I'd also like to understand the advantages of structuring the app using a Swift package versus adding frameworks to the main project. If there are resources (examples or explanations) you can point me to that would help me understand how this stuff works, that'd be great too. Thanks!

**Comment by mbrandonw** [comment, marked as answer]
Hey @gerfarfel,

This is what the root directory of the project looks like right now:

![image](https://user-images.githubusercontent.com/135203/107398056-1db52380-6ab4-11eb-8a43-b300f933495d.png)

It's a regular swift package at the root, and then the actual iOS application is in the `client` directory and it has the Xcode project. Also at the root we have a Xcode workspace and we just drag in the client Xcode project and the `Package.swift` file, and Xcode figures out all the rest. You can build the any of the SPM modules in isolation, and you can depend on any of the SPM modules from the iOS client.

One trick is you'll want to put a `Package.swift` in the `client` directory so that SPM doesn't try to load those files. The file can just hold this:

```swift
// swift-tools-version:5.2
import PackageDescription
let package = Package(name: "client", products: [], targets: [])
```

When all that is set up all you need to do to add a new module is add a few lines to `Package.swift`, create a directory in `Sources/` and then add code. It's very simple.

We use this style instead of frameworks in the Xcode project because it is very lightweight to add new modules and to describe the dependencies between modules. It's all captured in the `Package.swift` file.

Does that answer your questions?

**Reply by gerfarfel** [reply]
That's very helpful, thanks! I think I'm pretty close to reproducing your set-up, but I'm missing one piece. It seems that you must have individual Xcode schemes for each module in the package, since you're able to build each one individually. Is that correct, and if so, when and how are those schemes generated?

**Reply by mbrandonw** [reply]
Huh yeah, usually those are created automatically by SPM, however on occasion I do have to create them manually. Fortunately you can just go to Product > Scheme > Manage Schemes and hit the "Autocreate Schemes Now" button. That should create them all for you.
